src/tools/rbd_mirror/LeaderWatcher.h

```diff
@@ -0,0 +1,201 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_RBD_MIRROR_LEADER_WATCHER_H
+#define CEPH_RBD_MIRROR_LEADER_WATCHER_H
+
+#include <list>
+#include <memory>
+#include <string>
+
+#include "librbd/ManagedLock.h"
+#include "librbd/managed_lock/Types.h"
+#include "librbd/Watcher.h"
+#include "MirrorStatusWatcher.h"
+#include "tools/rbd_mirror/leader_watcher/Types.h"
+
+namespace librbd { class ImageCtx; }
+
+namespace rbd {
+namespace mirror {
+
+struct Threads;
+
+class LeaderWatcher : protected librbd::Watcher {
+public:
+  LeaderWatcher(Threads *threads, librados::IoCtx &io_ctx);
+
+  int init();
+  void shut_down();
+  bool is_leader();
+  void release_leader();
+
+  virtual void post_acquire_leader_handler(Context *on_finish) = 0;
```

Comment:
Perhaps just create a listener interface to avoid having to inherit from this class just to receive callbacks.